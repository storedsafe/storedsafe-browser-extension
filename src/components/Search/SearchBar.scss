@import '../../variables';

.search-bar {
  position: relative;
  display: flex;
  flex-direction: row;
  align-items: center;
  height: 100%;
  padding: $spacing;

  &.disabled {
    display: none;
  }

  .search-bar-input {
    flex-grow: 1;
    height: $menu-size;
    background: $color-primary-3;
    color: $color-fg-alt;
    border-bottom-color: $color-primary-4;
    box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1);

    &:focus {
      border-bottom-color: $color-accent-0;
    }
  }

  .search-bar-button {
    border: 0;
    outline: 0;
    background-color: $color-accent-0;
    cursor: pointer;
    height: $menu-size;
    width: $menu-size;
    border-radius: 0 $border-radius $border-radius 0;
    padding: $spacing;
    transition: $transition-duration;
    transition-property: background-color;
    box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1);

    &:hover,
    &:focus,
    &.selected {
      background-color: $color-bg;
    }

    > svg {
      height: 100%;
      width: 100%;
      transition: fill $transition-duration;
      fill: $color-fg-alt;
    }

    &:hover > svg,
    &:focus > svg,
    &.selected > svg {
      fill: $color-accent-0;
    }

    &.loading > svg {
      animation: loading 0.5s ease-in-out alternate infinite;
    }
  }
}

@keyframes loading {
  60% {
    transform: scale(1);
  }

  100% {
    transform: scale(1.1);
  }
}
