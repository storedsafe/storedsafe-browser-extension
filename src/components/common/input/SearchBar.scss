@import '../../../styles/variables.scss';

.search-bar {
  .search-form {
    width: 100%;
    height: 100%;
    display: flex;
    flex-direction: row;
    align-items: center;
  }

  &.disabled {
    display: none;
  }

  .search-bar-input {
    flex-grow: 1;
    height: $menu-size;
    background: $color-primary-3;
    color: $color-fg-alt;
    box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1);

    border-bottom-color: $color-primary-3;

    &:focus {
      border-bottom-color: $color-border-focus;
    }

    &:active {
      border-bottom-color: $color-accent-3;
    }
  }

  .search-button {
    background-color: $color-accent-0;
    border-bottom-left-radius: 0;
    height: $menu-size;
    width: $menu-size;
    transition: $transition-duration;

    &:hover,
    &:focus {
      background-color: $color-bg;
    }

    &:hover > svg,
    &:focus > svg {
      fill: $color-accent-0;
    }

    &.loading > svg {
      fill: transparent;
      stroke: $color-fg-alt;
      stroke-width: 0.5;
      stroke-dasharray: 32;
      animation: loading 2s ease-in-out alternate infinite;
    }

    &:hover.loading > svg,
    &:focus.loading > svg {
      fill: transparent;
      stroke: $color-accent-0;
    }
  }
}

@keyframes loading {
  0% {
    stroke-dashoffset: 0;
  }

  10% {
    stroke-dashoffset: 0;
  }

  90% {
    stroke-dashoffset: 64;
  }

  100% {
    stroke-dashoffset: 64;
  }
}