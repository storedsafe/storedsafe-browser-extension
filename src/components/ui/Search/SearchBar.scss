@import '../../../variables';

.search-bar {
  position: relative;
  display: flex;
  flex-direction: row;
  align-items: center;
  height: 100%;
  padding: $spacing;

  &.disabled {
    visibility: hidden;
  }

  .search-bar-input {
    flex-grow: 1;
    height: 64px - $spacing * 2;
    background: $color-primary-3;
    color: $color-fg-alt;
    border-bottom-color: $color-primary-4;
    box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1);

    &:focus {
      border-bottom-color: $color-accent-0;
    }
  }

  .search-bar-button {
    border: 0;
    outline: 0;
    background-color: $color-accent-0;
    cursor: pointer;
    height: $menu-size;
    width: $menu-size;
    border-radius: 0 $border-radius $border-radius 0;
    padding: $spacing;
    transition: $transition-duration;
    transition-property: background-color;
    box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1);

    &:hover,
    &:focus,
    &.selected {
      background-color: $color-bg;
    }

    > svg {
      height: 100%;
      width: 100%;
      transition: fill $transition-duration;
      fill: $color-fg-alt;
    }

    &:hover > svg,
    &:focus > svg,
    &.selected > svg {
      fill: $color-accent-0;
    }
  }
}
