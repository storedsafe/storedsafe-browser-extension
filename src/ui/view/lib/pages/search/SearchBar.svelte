<script lang="ts">
  import { createEventDispatcher } from "svelte";
  import { searchIcon } from "../../../../../global/icons";
  import { getMessage, LocalizedMessage } from "../../../../../global/i18n";

  import Icon from "../../layout/Icon.svelte";
  import {
    loading,
    messages,
    MessageType,
    search,
    SEARCH_LOADING_ID,
  } from "../../../../stores";

  // Submit search form automatically after `searchDelay` ms.
  export let searchDelay: number = 500;
  // Automatically focus search input
  export let focus: boolean = true;

  // Reference to form height for button size adjustment
  let formHeight: number;
  // Reference to timer to perform automatic search
  let timerId: number = null;
  // Set a fallback default button size until input reference is loaded
  const DEFAULT_BUTTON_SIZE = 36;
  // Search term (needle to find in the StoredSafe haystack)
  let needle: string = "";
  // Set up icon for search button
  const searchIconProps = {
    color: "#ebeef0",
    size: "1.2em",
    d: searchIcon,
  };

  /**
   * Perform search automatically after `searchDelay` ms since last change.
   * Will reset the timeout if search string is changed before the timeout is reached.
   **/
  function handleChange() {
    if (timerId) window.clearTimeout(timerId);
    timerId = window.setTimeout(find, searchDelay);
  }

  /**
   * Dispatch a search event when the search form is submitted.
   * */
  function handleSearch() {
    find();
  }

  function find(): void {
    loading.add(`${SEARCH_LOADING_ID}.${needle}`, search.search(needle), {
      onSuccess() {
        messages.clear();
      },
      onError(error) {
        messages.add(error.message, MessageType.ERROR);
      },
    });
  }
</script>

<!--
  @component
  Search form component, intended for putting at the top of the page.
  Will span the width of its parent and match the height of its contents.
-->
<form
  class="grid"
  style={`--button-size: ${formHeight ?? DEFAULT_BUTTON_SIZE}px`}
  role="search"
  bind:offsetHeight={formHeight}
  on:submit|preventDefault={handleSearch}
>
  <!-- svelte-ignore a11y-autofocus -->
  <input
    on:focus
    on:input={handleChange}
    bind:value={needle}
    autofocus={focus}
    placeholder={getMessage(LocalizedMessage.SEARCH_PLACEHOLDER)}
    aria-label={getMessage(LocalizedMessage.SEARCH_ARIA_LABEL_NEEDLE)}
    type="search"
  />
  <button
    type="submit"
    aria-label={getMessage(LocalizedMessage.SEARCH_ARIA_LABEL_SUBMIT)}
  >
    <Icon {...searchIconProps} />
  </button>
</form>

<style>
  form {
    grid-template-columns: 1fr var(--button-size);
    width: 100%;
    box-sizing: border-box;
  }

  input {
    background: var(--color-primary-dark);
    border-color: var(--color-primary-dark);
    color: #fff;
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
    border-right: 0;
  }

  button {
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
    border-left: 0;
  }

  input,
  button {
    margin: 0;
  }

  input:not(:disabled):hover,
  input:not(:disabled):focus {
    background-color: var(--color-primary-dark);
  }
</style>
