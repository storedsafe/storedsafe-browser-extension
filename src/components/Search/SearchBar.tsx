import React from 'react';
import svg from '../../ico/svg';
import './SearchBar.scss';

export type OnNeedleChangeCallback = (needle: string) => void;
export type OnSearchCallback = () => void;

export interface SearchBarProps {
  needle: string;
  onFocus?: () => void;
  onBlur?: () => void;
  onNeedleChange: OnNeedleChangeCallback;
  onSearch: OnSearchCallback;
  disabled?: boolean;
  isLoading: boolean;
}

export const SearchBar: React.FunctionComponent<SearchBarProps> = ({
  needle,
  onNeedleChange,
  onSearch,
  onFocus,
  onBlur,
  disabled,
  isLoading,
}: SearchBarProps) => {
  const onSubmit = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();
    onSearch();
  };

  return (
    <form className={`search-bar${disabled ? ' disabled' : ''}`} onSubmit={onSubmit}>
      <input
        className="search-bar-input"
        type="search"
        value={needle}
        placeholder="Search"
        aria-label="Search Text"
        disabled={disabled}
        onChange={({ target }: React.ChangeEvent<HTMLInputElement>): void => {
          onNeedleChange(target.value)
        }}
        onFocus={(): void => onFocus && onFocus()}
        onBlur={(): void => onBlur && onBlur()}
      />
      <button
        className={`search-bar-button${isLoading ? ' loading' : ''}`}
        type="submit"
        aria-label="Search Submit"
        disabled={disabled}>
        {svg.search}
      </button>
    </form>
  );
}

SearchBar.defaultProps = {
  disabled: false,
};
