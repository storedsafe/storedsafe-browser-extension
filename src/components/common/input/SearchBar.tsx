import React from 'react'
import './SearchBar.scss'
import { MenuButton } from './MenuButton'

const icons = {
  search: (
    <svg
      width='40'
      height='40'
      version='1.1'
      viewBox='0 0 10.583 10.583'
      xmlns='http://www.w3.org/2000/svg'
    >
      <path d='m7.1231.39901a3.0579 3.0579 0 00-3.0546 3.0579 3.0579 3.0579 0 00.56619 1.7722 3.0579 3.0579 0 000 .0004778l-4.2357 4.2352.71957.71956 4.2352-4.2352a3.0579 3.0579 0 001.7726.56571 3.0579 3.0579 0 003.0579-3.0579 3.0579 3.0579 0 00-3.0579-3.0579 3.0579 3.0579 0 00-.00334 0zm-.029146.89635a2.1616 2.1616 0 01.03249 0 2.1616 2.1616 0 012.1616 2.1616 2.1616 2.1616 0 01-2.1616 2.1616 2.1616 2.1616 0 01-2.1616-2.1616 2.1616 2.1616 0 012.1291-2.1616z' />
    </svg>
  )
}

type SearchInputProps = React.DetailedHTMLProps<
  React.InputHTMLAttributes<HTMLInputElement>,
  HTMLInputElement
>

export type OnNeedleChangeCallback = (needle: string) => void
export type OnSearchCallback = () => void

export interface SearchBarProps extends SearchInputProps {
  needle: string
  onNeedleChange: OnNeedleChangeCallback
  onSearch: OnSearchCallback
  disabled?: boolean
  isLoading: boolean
}

export const SearchBar: React.FunctionComponent<SearchBarProps> = ({
  needle,
  onNeedleChange,
  onSearch,
  disabled,
  isLoading,
  ...props
}: SearchBarProps) => {
  const onSubmit = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault()
    onSearch()
  }
  return (
    <section className='search-bar'>
      <form
        className={`search-form${disabled ? ' disabled' : ''}`}
        onSubmit={onSubmit}
      >
        <input
          placeholder='Search'
          title='Search'
          aria-label='Search Text'
          {...props}
          className='search-bar-input'
          type='search'
          value={needle}
          disabled={disabled}
          onChange={({ target }: React.ChangeEvent<HTMLInputElement>): void => {
            onNeedleChange(target.value)
          }}
        />
        <MenuButton
          className={`search-button${
            isLoading ? ' loading' : ''
          }`}
          type='submit'
          aria-label='Search Submit'
          title='Search'
          disabled={disabled}
          icon={icons.search}
        />
      </form>
    </section>
  )
}

SearchBar.defaultProps = {
  disabled: false
}
