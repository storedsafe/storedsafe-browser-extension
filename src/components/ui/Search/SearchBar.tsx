import React from 'react';
import svg from '../../../ico/svg';
import './SearchBar.scss';

export interface SearchBarProps {
  needle: string;
  onFocus?: () => void;
  onBlur?: () => void;
  onChange: (needle: string) => void;
  onSearch: (needle: string) => void;
  disabled?: boolean;
}

export const SearchBar: React.FunctionComponent<SearchBarProps> = ({
  needle,
  onChange,
  onSearch,
  onFocus,
  onBlur,
  disabled,
}: SearchBarProps) => {
  const onSubmit = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();
    onSearch(needle);
  };

  return (
    <form className={`search-bar${disabled ? ' disabled' : ''}`} onSubmit={onSubmit}>
      <input
        className="search-bar-input"
        type="search"
        value={needle}
        placeholder="Search"
        aria-label="Search Text"
        disabled={disabled}
        onChange={({ target }: React.ChangeEvent<HTMLInputElement>): void => {
          onChange(target.value)
        }}
        onFocus={(): void => onFocus && onFocus()}
        onBlur={(): void => onBlur && onBlur()}
      />
      <button className="search-bar-button" type="submit" aria-label="Search Submit" disabled={disabled}>
        {svg.search}
      </button>
    </form>
  );
}

SearchBar.defaultProps = {
  disabled: false,
};
