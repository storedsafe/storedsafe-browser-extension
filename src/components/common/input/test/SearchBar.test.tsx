import * as React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'
import { act, Simulate } from 'react-dom/test-utils'
import { SearchBar } from '../SearchBar'
import pretty from 'pretty'

/**
 * Set up and tear down container to mount tested component in.
 * */
let container: Element = null
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div')
  document.body.appendChild(container)
})
afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container)
  container.remove()
  container = null
})

test('<SearchBar isLoading=true />', () => {
  const needle = 'foo'
  const isLoading = true
  const onSearch = jest.fn()
  const onNeedleChange = jest.fn()

  act(() => {
    render(
      <SearchBar
        needle={needle}
        onSearch={onSearch}
        onNeedleChange={onNeedleChange}
        isLoading={isLoading}
      />,
      container
    )
  })

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test('<SearchBar isLoading=false />', () => {
  const needle = 'foo'
  const isLoading = false
  const onSearch = jest.fn()
  const onNeedleChange = jest.fn()

  act(() => {
    render(
      <SearchBar
        needle={needle}
        onSearch={onSearch}
        onNeedleChange={onNeedleChange}
        isLoading={isLoading}
      />,
      container
    )
  })
})

test('<SearchBar disabled />', () => {
  const needle = 'foo'
  const isLoading = false
  const disabled = true
  const onSearch = jest.fn()
  const onNeedleChange = jest.fn()

  act(() => {
    render(
      <SearchBar
        needle={needle}
        onSearch={onSearch}
        onNeedleChange={onNeedleChange}
        isLoading={isLoading}
        disabled={disabled}
      />,
      container
    )
  })
})

test('<SearchBar interaction />', () => {
  const needle = 'foo'
  const isLoading = false
  const onSearch = jest.fn()
  const onNeedleChange = jest.fn()

  act(() => {
    render(
      <SearchBar
        needle={needle}
        onSearch={onSearch}
        onNeedleChange={onNeedleChange}
        isLoading={isLoading}
      />,
      container
    )
  })

  const searchInput: HTMLInputElement = container.querySelector(
    '.search-bar-input'
  )
  act(() => {
    searchInput.value = 'bar'
    Simulate.change(searchInput)
  })
  expect(onNeedleChange).toHaveBeenCalledWith('bar')

  act(() => {
    Simulate.submit(searchInput)
  })
  expect(onSearch).toHaveBeenCalledWith()
})
