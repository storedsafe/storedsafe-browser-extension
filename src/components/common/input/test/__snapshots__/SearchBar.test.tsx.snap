// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar isLoading=true /> 1`] = `
"<section class=\\"search-bar\\">
  <form class=\\"search-form\\"><input placeholder=\\"Search\\" title=\\"Search\\" aria-label=\\"Search Text\\" class=\\"search-bar-input\\" type=\\"search\\" value=\\"foo\\"><button class=\\"menu-button search-button loading\\" type=\\"submit\\" aria-label=\\"Search Submit\\" title=\\"Search\\"><svg width=\\"40\\" height=\\"40\\" version=\\"1.1\\" viewBox=\\"0 0 10.583 10.583\\" xmlns=\\"http://www.w3.org/2000/svg\\">
        <path d=\\"m7.1231.39901a3.0579 3.0579 0 00-3.0546 3.0579 3.0579 3.0579 0 00.56619 1.7722 3.0579 3.0579 0 000 .0004778l-4.2357 4.2352.71957.71956 4.2352-4.2352a3.0579 3.0579 0 001.7726.56571 3.0579 3.0579 0 003.0579-3.0579 3.0579 3.0579 0 00-3.0579-3.0579 3.0579 3.0579 0 00-.00334 0zm-.029146.89635a2.1616 2.1616 0 01.03249 0 2.1616 2.1616 0 012.1616 2.1616 2.1616 2.1616 0 01-2.1616 2.1616 2.1616 2.1616 0 01-2.1616-2.1616 2.1616 2.1616 0 012.1291-2.1616z\\"></path>
      </svg></button></form>
</section>"
`;
